### PR TITLE
cmake,crc32c: conditionalize crc32c on different archs

### PR DESCRIPTION
**Pull request by tchaikov**

### CODE REVIEW

**Comment by wjwithagen** [comment]
@tchaikov 
I see at least `bin/unittest_crc32c` complain about missing ARM CRC.

Perhaps solve all the include stuff in `include/crc32c.h`

**Comment by tchaikov** [comment]
@wjwithagen fixed and repushed.

**Comment by wjwithagen** [comment]
@tchaikov 
LGTM

**Comment by tchaikov** [comment]
@kestrels could you help take a look at this change also?

**Comment by kestrels** [comment]
Sure.

**Comment by kestrels** [comment]
This is hitting a build break for me on my ppc64le system.  The break is coming from src/common/simple_spin.cc.  

[ 40%] Building CXX object src/librbd/CMakeFiles/rbd_internal.dir/image/CreateRequest.cc.o
[ 40%] Building CXX object src/CMakeFiles/common-objs.dir/common/code_environment.cc.o
[ 40%] Building CXX object src/CMakeFiles/common-objs.dir/common/dout.cc.o
[ 40%] Building CXX object src/CMakeFiles/common-objs.dir/common/signal.cc.o
[ 40%] Building CXX object src/CMakeFiles/common-objs.dir/common/simple_spin.cc.o
/home/ubuntu/ceph/tchaikov.1/ceph/src/common/simple_spin.cc:38:2: error: #error "Unknown architecture"
 #error "Unknown architecture"
  ^
src/CMakeFiles/common-objs.dir/build.make:2415: recipe for target 'src/CMakeFiles/common-objs.dir/common/simple_spin.cc.o' failed
make[2]: *** [src/CMakeFiles/common-objs.dir/common/simple_spin.cc.o] Error 1
CMakeFiles/Makefile2:686: recipe for target 'src/CMakeFiles/common-objs.dir/all' failed
make[1]: *** [src/CMakeFiles/common-objs.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 40%] Building CXX object src/mds/CMakeFiles/mds.dir/CDir.cc.o
[ 40%] Building CXX object src/librbd/CMakeFiles/rbd_internal.dir/image/OpenRequest.cc.o
[ 40%] Building CXX object src/mds/CMakeFiles/mds.dir/CInode.cc.o
[ 40%] Building CXX object src/mds/CMakeFiles/mds.dir/LogEvent.cc.o

The __ppc64__ is not defined.  It should probably be changed to __powerpc__ or __powerpc64__, both of which are defined.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -527,9 +510,16 @@ set(libcommon_files
...
   common/dns_resolve.cc
   common/hostname.cc
-  ${arch_files}
+  arch/probe.cc
   ${auth_files}
   ${mds_files})
+if(HAVE_ARM)
+  list(APPEND libcommon_files arch/arm.c)
+elseif(HAVE_INTEL)
+  list(APPEND libcommon_files arch/intel.c)
```
Comment (kestrels) [inline src/CMakeLists.txt]: There are symbols defined in src/arch/intel.c (e.g. ceph_arch_intel_pclmul and ceph_arch_intel_sse41) that are referenced in other parts of the code outside of any ifdef.  

I think the answer is probably to either (a) go back to unconditionally having arch/intel.c unconditionally (even on non-intel compiles) or (b) go forward with finding all references to those symbols and hiding them behind the proper intel related ifdefs.  

[100%] ?[32m?[1mLinking CXX executable ../../bin/radosgw-admin?[0m
[100%] Built target radosgw-admin
../../../lib/libcephd.a(libcephd.cc.o):(.toc+0x10): undefined reference to `ceph_arch_intel_pclmul'
../../../lib/libcephd.a(libcephd.cc.o):(.toc+0x18): undefined reference to `ceph_arch_intel_sse41'
collect2: error: ld returned 1 exit status
src/test/libcephd/CMakeFiles/ceph_test_cephd_api_misc.dir/build.make:147: recipe for target 'bin/ceph_test_cephd_api_misc' failed
make[2]: *** [bin/ceph_test_cephd_api_misc] Error 1
CMakeFiles/Makefile2:13507: recipe for target 'src/test/libcephd/CMakeFiles/ceph_test_cephd_api_misc.dir/all' failed
make[1]: *** [src/test/libcephd/CMakeFiles/ceph_test_cephd_api_misc.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[100%] ?[32m?[1mLinking CXX executable ../bin/ceph-dencoder?[0m
[100%] Built target ceph-dencoder

~/ceph/tchaikov.1/ceph/src/compressor/zlib$ cat CompressionPluginZlib.h | grep intel
#include "arch/intel.h"
      isal = (ceph_arch_intel_pclmul && ceph_arch_intel_sse41);
~/ceph/tchaikov.1/ceph/src/compressor/zlib$

**Comment by tchaikov** [comment]
@kestrels the second build failure is addressed and the fix is re-pushed.

**Comment by kestrels** [comment]
With both #14289 and #14310, the build completed successfully.  The unittest_crc32c confirmed that the ppc specific crc algorithm is being used.

**Pull request merged**